src/interface.jl

```diff
@@ -0,0 +1,655 @@
+implement `jacobian` and, by impliciation, `logabsdetjac`.
+"""
+abstract type ADBijector{AD} <: Bijector end
+
+"""
+    inv(b::Bijector)
+    Inversed(b::Bijector)
+
+A `Bijector` representing the inverse transform of `b`.
+"""
+struct Inversed{B <: Bijector} <: Bijector
+    orig::B
+end
+
+inv(b::Bijector) = Inversed(b)
+inv(ib::Inversed{<:Bijector}) = ib.orig
+
+"""
+    logabsdetjac(b::Bijector, x)
+    logabsdetjac(ib::Inversed{<: Bijector}, y)
+
+Computes the log(abs(det(J(x)))) where J is the jacobian of the transform.
+Similarily for the inverse-transform.
+
+Default implementation for `Inversed{<: Bijector}` is implemented as
+`- logabsdetjac` of original `Bijector`.
+"""
+logabsdetjac(ib::Inversed{<: Bijector}, y) = - logabsdetjac(ib.orig, ib(y))
+
+"""
+    forward(b::Bijector, x)
+    forward(ib::Inversed{<: Bijector}, y)
+
+Computes both `transform` and `logabsdetjac` in one forward pass, and
+returns a named tuple `(rv=b(x), logabsdetjac=logabsdetjac(b, x))`.
+
+This defaults to the call above, but often one can re-use computation
+in the computation of the forward pass and the computation of the
+`logabsdetjac`. `forward` allows the user to take advantange of such
+efficiencies, if they exist.
+"""
+forward(b::Bijector, x) = (rv=b(x), logabsdetjac=logabsdetjac(b, x))
+forward(ib::Inversed{<: Bijector}, y) = (
+    rv=ib(y),
+    logabsdetjac=logabsdetjac(ib, y)
+)
+
+
+# AD implementations
+function jacobian(b::ADBijector{<: ForwardDiffAD}, x::Real)
+    return ForwardDiff.derivative(b, x)
+end
+function jacobian(b::Inversed{<: ADBijector{<: ForwardDiffAD}}, y::Real)
+    return ForwardDiff.derivative(b, y)
+end
+function jacobian(b::ADBijector{<: ForwardDiffAD}, x::AbstractVector{<: Real})
+    return ForwardDiff.jacobian(b, x)
+end
+function jacobian(b::Inversed{<: ADBijector{<: ForwardDiffAD}}, y::AbstractVector{<: Real})
+    return ForwardDiff.jacobian(b, y)
+end
+
+function jacobian(b::ADBijector{<: TrackerAD}, x::Real)
+    return Tracker.gradient(b, x)[1]
+end
+function jacobian(b::Inversed{<: ADBijector{<: TrackerAD}}, y::Real)
+    return Tracker.gradient(b, y)[1]
+end
+function jacobian(b::ADBijector{<: TrackerAD}, x::AbstractVector{<: Real})
+    # we extract `data` so that we don't returne a `Tracked` type
+    return Tracker.data(Tracker.jacobian(b, x))
+end
+function jacobian(b::Inversed{<: ADBijector{<: TrackerAD}}, y::AbstractVector{<: Real})
+    # we extract `data` so that we don't returne a `Tracked` type
+    return Tracker.data(Tracker.jacobian(b, y))
+end
+
+# TODO: allow batch-computation, especially for univariate case?
+"Computes the absolute determinant of the Jacobian of the inverse-transformation."
+logabsdetjac(b::ADBijector, x::Real) = log(abs(jacobian(b, x)))
+function logabsdetjac(b::ADBijector, x::AbstractVector{<:Real})
+    fact = lu(jacobian(b, x), check=false)
+    return issuccess(fact) ? log(abs(det(fact))) : -Inf # TODO: do this or not?
```

Comment:
Yeah, I didn't conclude with anything so I just left it as-is. But sure, I'm okay with throwing a warning. Should we make it a "custom" warning or just call `logabsdet` and let it throw the `SingularException`?
I think I like a custom exception so that the user fully understands that the Jacobian is for some reason singular. Just getting a `SingularException` takes a bit more of a mental effort to track down

Reply:
How about
```julia
struct SingularJacobianException{B} <: Exception where B <: Bijector
    b::B
end
Base.showerror(io::IO, e::SingularJacobianException) = print(io, "jacobian of $(e.b) is singular")

# TODO: allow batch-computation, especially for univariate case?
"Computes the absolute determinant of the Jacobian of the inverse-transformation."
function logabsdetjac(b::ADBijector, x::Real)
    res = log(abs(jacobian(b, x)))
    return isfinite(res) ? res : throw(SingularJacobianException(b))
end

function logabsdetjac(b::ADBijector, x::AbstractVector{<:Real})
    fact = lu(jacobian(b, x), check=false)
    return issuccess(fact) ? log(abs(det(fact))) : throw(SingularJacobianException(b))
end
```
Then
```julia
julia> struct NonInvertibleBijector{AD} <: ADBijector{AD} end

julia> (b::NonInvertibleBijector)(x) = clamp.(x, 0, 1)

julia> b = NonInvertibleBijector{Bijectors.ADBackend()}()
NonInvertibleBijector{Bijectors.ForwardDiffAD}()

julia> Bijectors.jacobian(b, [1.0, 10.0])
2×2 Array{Float64,2}:
 1.0  0.0
 0.0  0.0

julia> logabsdetjac(b, [1.0, 10.0])
ERROR: jacobian of NonInvertibleBijector{Bijectors.ForwardDiffAD}() is singular
Stacktrace:
 [1] logabsdetjac(::NonInvertibleBijector{Bijectors.ForwardDiffAD}, ::Array{Float64,1}) at /home/tor/.julia/dev/Bijectors/src/interface.jl:134
 [2] top-level scope at none:0
```

It's worth pointing out that such a check will not be performed for cases where `logabsdetjac` does _not_ use AD.

Reply:
An exception works too. I don't have strong opinion on this.

PS: I originall was suggesting to keep returning `-Inf` but in the meantime thorwing a warning message not an exception.